accepted/2021/winforms/streamline-application-bootstrap.md

```diff
@@ -0,0 +1,362 @@
+1. Use Roslyn analyzers in Visual Basic scenarios until previous two items are addressed.
+
+
+:thinking: Strategically it could be benefitial to migrate Visual Basic off the Application Framework and *.myapp file in favour of Roslyn source generators and MSBuild properties. This could also significantly reduce efforts in maintaining Visual Studio property pages for Visual Studio projects (e.g. [dotnet/project-system#7236](https://github.com/dotnet/project-system/issues/7236), [dotnet/project-system#7240](https://github.com/dotnet/project-system/issues/7240), and [dotnet/project-system#7241](https://github.com/dotnet/project-system/issues/7241))
+
+
+
+## Stakeholders and Reviewers
+
+<!--
+We noticed that even in the cases where we have specs, we sometimes surprise key
+stakeholders because we didn't pro-actively involve them in the initial reviews
+and early design process.
+
+Please take a moment and add a bullet point list of teams and individuals you
+think should be involved in the design process and ensure they are involved
+(which might mean being tagged on GitHub issues, invited to meetings, or sent
+early drafts).
+-->
+
+* Windows Forms runtime/designer: @dotnet/dotnet-winforms @OliaG @DustinCampbell 
+* Visual Basic: @KathleenDollard  @KlausLoeffelmann
+* Roslyn: @jaredpar @sharwell
+* Project System: @drewnoakes 
+* SDK: dsplaisted 
+* .NET libraries: @ericstj
+* General: @terrajobst @Pilchie 
+
+## Design
+
+<!--
+This section will likely have various subheadings. The structure is completely
+up to you and your engineering team. It doesn't need to be complete; the goal is
+to provide enough information so that the engineering team can build the
+feature.
+
+If you're building an API, you should include the API surface, for example
+assembly names, type names, method signatures etc. If you're building command
+line tools, you likely want to list all commands and options. If you're building
+UI, you likely want to show the screens and intended flow.
+
+In many cases embedding the information here might not be viable because the
+document format isn't text (for instance, because it's an Excel document or in a
+PowerPoint deck). Add links here. Ideally, those documents live next to this
+document.
+-->
+
+
+### 1. Visual Basic 
+
+Ironically Visual Basic apps are in a good position for the designer integration with their Application Framework functionality, i.e. "*.myapp" file:
+
+![image](https://user-images.githubusercontent.com/4403806/114354200-3d9abd80-9bb1-11eb-9d48-a43f299f82e1.png)
+
+The designer surface process should have no issues reading configuration values from a myapp file. Any work further work at this stage pertaining for streamlining the bootstrap is out of scope of this design proposal.
+
+
+
+### 2. C#
+
+```diff
+    static class Program
+    {
+        [STAThread]
+        static void Main()
+        {
++           ApplicationConfiguration.Initialize();
```

Comment:
It's unfortunate that this can't be `Application.Initialize()`.

Alternatively, we could follow the pattern that we have used in the JSON generator where the user controls the type. This could look as follows:


```C#
// Program.cs, written by the user

static partial class Program
{
    [STAThread]
    static void Main()
    {
        InitializeApplication();
        Application.Run(new MainForm());
    }

    [GenerateInitializeApplication]
    private partial void InitializeApplication();
}

// Program.g.cs, generated

partial class Program
{
    private void InitializeApplication()
    {
        Application.EnableVisualStyles();
        Application.SetDefaultFont(new Font(....));
        Application.SetHighDpiMode(HighDpiMode.PerMonitorV2);
        Application.SetCompatibleTextRenderingDefault(false);
    }
}
```

We could also decide that we don't need the user to opt-in at the source level and decide that setting the MSBuild properties is sufficient, which would reduce it to just this:

```C#
// Program.cs, written by the user

static partial class Program
{
    [STAThread]
    static void Main()
    {
        InitializeApplication();
        Application.Run(new MainForm());
    }
}

// Program.g.cs, generated

partial class Program
{
    private void InitializeApplication()
    {
        Application.EnableVisualStyles();
        Application.SetDefaultFont(new Font(....));
        Application.SetHighDpiMode(HighDpiMode.PerMonitorV2);
        Application.SetCompatibleTextRenderingDefault(false);
    }
}
```

Reply:
> It's unfortunate that this can't be `Application.Initialize()`.

Yep.

> Alternatively, we could follow the pattern that we have used in the JSON generator where the user controls the type. This could look as follows: ...
> We could also decide that we don't need the user to opt-in at the source level and decide that setting the MSBuild properties is sufficient, which would reduce it to just this: ...

I have no strong preference either way.

Reply:
Ah Actually I have a better idea:
- ``Application.RegisterMainForm(Form form)`` (could be added to winforms)
- ``Application.Run()`` (no args) could be made to check if a main form is registered then do as if ``Application.Run(Form form)`` was called.
- The generated ``Initialize()`` method calls ``Application.Run()`` and leaves the developer to call ``Application.RegisterMainForm(Form form)`` (which then stores it for use in ``Application.Run()`` as a variable that is then checked for null in the implementation of ``Application.Run()``)

In theory it should work, but in general it could be a breaking change for .NET 6 (not to mention possibly too late to propose ``Application.RegisterMainForm(Form form)``) and having the generator call ``Application.Run()`` for the developer inside of the generated code.

99.9999999% of cases there are only a single main form anyway with the main form spinning up the child forms directly at startup of that main form.

Reply:
@terrajobst if we're going for top level statements I don't think the partial `Program` will work:
![image](https://user-images.githubusercontent.com/4403806/121961464-cf3dcb80-cdaa-11eb-80ad-bd0ce8964312.png)